### PR TITLE
TINKERPOP-2978: Implement list functions.

### PR DESCRIPTION
**Pull request by kenhuuu**
Includes combine, difference, disjunct, intersect, conjoin, merge, product and reverse.

Jira: https://issues.apache.org/jira/browse/TINKERPOP-2978

### CODE REVIEW

File: docs/src/reference/the-traversal.asciidoc [hunk 1]
```diff
@@ -3759,12 +3885,14 @@ link:++https://tinkerpop.apache.org/javadocs/x.y.z/core/org/apache/tinkerpop/gre
...
-The `reverse()`-step (*map*) returns a reversed string of the incoming string traverser. Null values are not processed and remain
-as null when returned. If the incoming traverser is a non-String value then an `IllegalArgumentException` will be thrown.
+The `reverse()`-step (*map*) returns a reversed of the incoming list traverser. Single values (including null) are not
+processed and are added back to the Traversal Stream unchanged. If the incoming traverser is a String value then the
+reversed String will be returned.
 
 [gremlin-groovy,modern]
 ----
 g.V().values("name").reverse()
+g.V().values("name").fold().reverse()
```
Comment (vkagamlyk) [inline docs/src/reference/the-traversal.asciidoc:3895]: good to add <1>/<2> and explain difference

File: docs/src/upgrade/release-3.7.x.asciidoc [hunk 2]
```diff
@@ -225,11 +226,34 @@ gremlin> g.V().values("age").fold().all(P.gt(10))
...
+==>[29,27,32,35,32,30,50]
+gremlin> g.V().values("age").fold().intersect([32,30,50])
+==>[32]
+gremlin> g.V().values("age").fold().disjunct([32,30,50])
+==>[50,35,27,29,30]
+gremlin> g.V().values("age").fold().difference([32,30,50])
+==>[35,27,29]
+gremlin> g.V().values("age").order().by(desc).fold().reverse()
+==>[27,29,32,35]
+gremlin> g.V().values("age").fold().join("-")
```
Comment (vkagamlyk) [inline docs/src/upgrade/release-3.7.x.asciidoc:243]: `join` or `conjoin`?

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConjoinStep.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A map step that returns the result of joining every element in the traverser using the delimiter argument.
+ */
+public final class ConjoinStep<S> extends ScalarMapStep<S, String> implements ListFunction {
```
Comment (vkagamlyk) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConjoinStep.java:36]: do we need to override `toString()` for some steps?

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/TraversalMergeStep.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            parameterItems = values;
+        }
+    }
+
+    @Override
+    public String getStepName() { return "merge"; }
+
+    @Override
+    protected E map(final Traverser.Admin<S> traverser) {
+        S incoming = traverser.get();
```
Comment (vkagamlyk) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/TraversalMergeStep.java:64]: nit: `final`

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/TraversalMergeStep.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        super.setTraversal(parentTraversal);
+        if (valueTraversal != null) { this.integrateChild(this.valueTraversal); }
+    }
+
+    @Override
+    public TraversalMergeStep<S, E> clone() {
+        final TraversalMergeStep<S, E> clone = (TraversalMergeStep<S, E>) super.clone();
+        if (null != this.valueTraversal) {
+            clone.valueTraversal = this.valueTraversal.clone();
+        } else if (this.parameterItems instanceof Map) {
```
Comment (vkagamlyk) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/TraversalMergeStep.java:119]: Why different handling depends on type of `parameterItems`?

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/CombineStep.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        return combined;
+    }
+
+    @Override
+    public List<Traversal.Admin<S, E>> getLocalChildren() {
+        return (null == valueTraversal) ? Collections.emptyList() : Collections.singletonList(valueTraversal);
+    }
+
+    @Override
+    public Set<TraverserRequirement> getRequirements() { return this.getSelfAndChildRequirements(); }
```
Comment (xiazcy) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/CombineStep.java:72]: I think you'd want to pass in the requirement for the self too?
```suggestion
    public Set<TraverserRequirement> getRequirements() { return this.getSelfAndChildRequirements(TraverserRequirement.OBJECT); }
```

File: docs/src/reference/the-traversal.asciidoc (hunk 1 above)
Comment (kenhuuu) [reply docs/src/reference/the-traversal.asciidoc]: Done.

File: docs/src/upgrade/release-3.7.x.asciidoc (hunk 2 above)
Comment (kenhuuu) [reply docs/src/upgrade/release-3.7.x.asciidoc]: conjoin. updated.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java [hunk 7]
```diff
@@ -1672,6 +1678,90 @@ public default GraphTraversal<S, Long> dateDiff(final Traversal<?, Date> dateTra
...
+    public default GraphTraversal<S, Set<?>> intersect(final Object values) {
+        this.asAdmin().getBytecode().addStep(Symbols.intersect, values);
+        return this.asAdmin().addStep(new IntersectStep<>(this.asAdmin(), values));
+    }
+
+    /**
+     * Joins together the elements of the incoming list traverser together with the provided delimiter.
+     *
+     * @return the traversal with an appended {@link ConjoinStep}.
+     * @see <a href="http://tinkerpop.apache.org/docs/${project.version}/reference/#join-step" target="_blank">Reference Documentation - Join Step</a>
```
[the remark this answers is not in the document]
Comment (kenhuuu) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java]: updated.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConjoinStep.java (hunk 3 above)
Comment (kenhuuu) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConjoinStep.java:36]: Maybe, I decided it wouldn't be that useful given that half of these fields are meant to be null. By default the step's name is already included in the toString which is probably the most important part.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/TraversalMergeStep.java (hunk 4 above)
Comment (kenhuuu) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/TraversalMergeStep.java]: updated.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/TraversalMergeStep.java (hunk 5 above)
Comment (kenhuuu) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/TraversalMergeStep.java]: This was probably a mistake. Updated.

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConjoinStepTest.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+import static org.junit.Assert.assertTrue;
+
+public class ConjoinStepTest extends StepTest {
+    @Override
+    protected List<Traversal> getTraversals() { return Collections.singletonList(__.conjoin("a")); }
+
+    @Test
+    public void testReturnTypes() {
+        try {
+            __.__(Collections.emptyList()).conjoin(null).next();
```
[the remark this answers is not in the document]
Comment (kenhuuu) [reply gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConjoinStepTest.java:42]: updated.

File: docs/src/dev/provider/gremlin-semantics.asciidoc [hunk 9]
```diff
@@ -725,6 +725,42 @@ link:https://github.com/apache/tinkerpop/tree/x.y.z/gremlin-core/src/main/java/o
...
+
+[width="100%",options="header"]
+|=========================================================
+|Start Step |Mid Step |Modulated |Domain |Range
+|N |Y |N |`array`/`Iterable` |`List`
+|=========================================================
+
+*Arguments:*
+
+* `Object` - A list of items (as an Iterable or an array) a traversal that will produce a list of items.
```
[the remark this answers is not in the document]
Comment (kenhuuu) [reply docs/src/dev/provider/gremlin-semantics.asciidoc]: updated.

File: docs/src/dev/provider/gremlin-semantics.asciidoc [hunk 10]
```diff
@@ -725,6 +725,42 @@ link:https://github.com/apache/tinkerpop/tree/x.y.z/gremlin-core/src/main/java/o
...
+will cause exceptions to be thrown.
+
+*Exceptions*
+
+* If the incoming traverser isn't a list (array or Iterable) then an `IllegalArgumentException` will be thrown.
+* If the argument doesn't resolve to a list (array or Iterable) then an `IllegalArgumentException` will be thrown.
+
+See: link:https://github.com/apache/tinkerpop/tree/x.y.z/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/CombineStep.java[source],
+link:https://tinkerpop.apache.org/docs/x.y.z/reference/#combine-step[reference],
+link:https://tinkerpop.apache.org/docs/x.y.z/reference/#merge-step[reference],
```
[the remark this answers is not in the document]
Comment (kenhuuu) [reply docs/src/dev/provider/gremlin-semantics.asciidoc]: yes it was intended. updated to your suugestion.

File: gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/Combine.feature [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+    And the traversal of
+      """
+      g.V().values("name").fold().combine(__.V().values("nonexistant").fold())
+      """
+    When iterated to list
+    Then the result should be unordered
+      | result |
+      | l[marko,vadas,lop,josh,ripple,peter] |
+
+  @GraphComputerVerificationInjectionNotSupported
```
[the remark this answers is not in the document]
Comment (kenhuuu) [reply gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/Combine.feature]: updated tag.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/CombineStep.java (hunk 6 above)
Comment (kenhuuu) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/CombineStep.java:72]: OBJECT is pretty much always required and so it doesn't need to be explicitly added here.

**Comment by vkagamlyk** [comment]
VOTE +1

File: docs/src/reference/the-traversal.asciidoc [hunk 12]
```diff
@@ -1140,6 +1140,24 @@ g.V().coin(1.0)
...
+=== Combine Step
+
+The `combine()`-step (*map*) combines the elements of the incoming list traverser and the provided list argument into
+one list. This is also known as appending or concatenating. This step only expects list data (array or Iterable) and
+will throw an `IllegalArgumentException` if any other type is encountered (including null). This differs from the
+`merge` step in that it allows duplicates to exist.
+
+[gremlin-groovy,modern]
+----
+g.V().values("name").fold().combine(["james","jen","marko","vadas"])
```
[the remark this answers is not in the document]
Comment (kenhuuu) [reply docs/src/reference/the-traversal.asciidoc:1153]: added to all steps.

**Comment by Cole-Greer** [comment]
Thanks Ken, these added steps will be a huge addition to gremlin's list manipulation capabilities.

VOTE +1

**Comment by kenhuuu** [comment]
VOTE +1

**Pull request merged**